lib/location/method_gnss.c

```diff
@@ -110,39 +159,45 @@ void method_gnss_timeout_work_fn(struct k_work *item)
 	method_gnss_cancel();
 }
 
-int method_gnss_init(void)
-{
-	int err = nrf_modem_gnss_init();
-
-	if (err) {
-		LOG_ERR("Failed to initialize GNSS interface, error %d", err);
-		return -err;
-	}
-
-	err = nrf_modem_gnss_event_handler_set(method_gnss_event_handler);
-	if (err) {
-		LOG_ERR("Failed to set GNSS event handler, error %d", err);
-		return -err;
-	}
-
-	k_work_init(&gnss_fix_work, method_gnss_fix_work_fn);
-	k_work_init(&method_gnss_timeout_work, method_gnss_timeout_work_fn);
-
-	return 0;
-}
-
-int method_gnss_location_get(const struct loc_method_config *config)
+static void method_gnss_positioning_work_fn(struct k_work *work)
 {
 	int err = 0;
-
-	const struct loc_gnss_config *gnss_config = &config->gnss;
+	int tau;
+	int active_time;
+	enum lte_lc_system_mode mode;
+
+	struct k_work_args *work_data =
+		CONTAINER_OF(work, struct k_work_args, work_item);
+
+	const struct loc_gnss_config gnss_config = work_data->gnss_config;
+
+	lte_lc_system_mode_get(&mode, NULL);
+
+	/* Don't care about PSM if we are in GNSS only mode */
+	if(mode != LTE_LC_SYSTEM_MODE_GPS) {
+		int ret = lte_lc_psm_get(&tau, &active_time);
+		if (ret < 0) {
+			LOG_ERR("Cannot get PSM config: %d. Starting GNSS right away.", ret);
+		} else if ((tau >= 0) & (active_time >= 0)) {
+			LOG_INF("Waiting for LTE to enter PSM: tau %d active_time %d",
+				tau, active_time);
+
+			/* Wait for the PSM to start. If semaphore is reset during the waiting
+			 * period, the position request was canceled. Thus, return without
+			 * doing anything */
+			if (k_sem_take(&entered_psm_mode, K_FOREVER) == -EAGAIN) {
```

Comment:
Gets stuck indefinitely if we never enter PSM.

Reply:
We can use a timeout here and I think we should. Need to discuss what would be correct approach.

Reply:
Yes, method timeout should handle this and there (in cancel handler of the method) semaphore should be released.

Reply:
I forgot to add that this waiting can nevertheless be interrupted by calling the cancel method.

Reply:
The timeout we planned in our previous meeting won't help here because it would start once we enter PSM and can start GNSS in practice. But yes, we'll need to keep this in mind and think how to solve it.

Reply:
We discussed that this would be changed to 5mins as a long enough time that will still prevent API getting stuck.